diff --git a/CMakeLists.txt b/CMakeLists.txt
index 1ed9fd3..e01043d 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -39,8 +39,8 @@ cmake_minimum_required(VERSION 3.5)
 
 project(aimet)
 
-set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -O0 -ggdb -fPIC -D_GLIBCXX_USE_CXX11_ABI=0")
-set(AIMET_INSTALL_DIR ${CMAKE_BINARY_DIR}/staging/universal)
+set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -O3 -fPIC -D_GLIBCXX_USE_CXX11_ABI=0")
+set(AIMET_INSTALL_DIR ${CMAKE_BINARY_DIR}/../install/staging/universal)
 
 if(NOT DEFINED AIMET_PYTHONPATH)
   set(AIMET_PYTHONPATH "PYTHONPATH=${CMAKE_BINARY_DIR}/artifacts" CACHE STRING "python path")
@@ -89,11 +89,10 @@ else(ENABLE_CUDA)
 
 endif(ENABLE_CUDA)
 
-set(LAPACKE_LIBRARY "/usr/lib/x86_64-linux-gnu/liblapacke.so.3.7.1")
-
+set(LAPACKE_LIBRARY "${LAPACKE_DIR}/lib/liblapacke.so")
+find_package (Eigen3 REQUIRED NO_MODULE)
 find_package(OpenCV REQUIRED)
-set(OPENCV_LIBRARY /usr/local/lib/libopencv_core.a)
-
+set(OPENCV_LIBRARY "${OpenCV_LIBS}")
 # ----------------------------------
 # Conditional build for PyTorch
 # ----------------------------------
